
<p align="center">
<img src="https://raw.githubusercontent.com/cgoliver/rnaglib/master/images/rgl.png#gh-light-mode-only" width="30%">
</p>


# RNA Geometric Library (`rnaglib`)
[![Documentation Status](https://readthedocs.org/projects/rnaglib/badge/?version=latest)](https://rnaglib.readthedocs.io/en/latest/?badge=latest)

`RNAglib` is a Python package for studying RNA 2.5D and 3D structures. Functionality includes automated data loading, analysis,
visualization, ML model building and benchmarking.

We host RNAs annotated with molecule, base pair, and nucleotide level attributes. These include, but are not limited to:

* Secondary structure
* 3D coordinates
* Protein binding 
* Small molecule binding 
* Chemical modifications 
* Leontis-westhof base pair geometry classification


![Example graph](https://raw.githubusercontent.com/cgoliver/rnaglib/master/images/rgl_fig.png)

## Installation

The package can be cloned and the source code used directly. We also deploy it as a pip package and recommend using this
install in conda environments.

If one wants to use GPU support, one should install [Pytorch](https://pytorch.org/get-started/locally/)
and [DGL](https://www.dgl.ai/pages/start.html) with the appropriate options. Otherwise you can just skip this step and
the pip installs of Pytorch and DGL will be used.

Then, one just needs to run :

```
pip install rnaglib
```

Then one can start using the packages functionalities by importing them in one's python script.


## What can you do with `rnaglib`?

### Benchmark ML models on RNA 3D structures (**new**)

Datasets of RNA 3D structures ready-to-use for machine learning model benchmarking in various biologically relevant tasks.

* One line to fetch annotated dataset and splits loading
* Encode structures as graphs, voxels and point clouds

All tasks take as input an RNA 3D structure and have as output either residue or whole RNA-level properties.

We currently support the following 6 prediction tasks and the ability to create [new tasks](https://rnaglib.readthedocs.io/en/latest/tasks.html#tutorial-2-creating-a-new-task):


* [Reidue-Level RNA-Protein Binding](https://github.com/cgoliver/rnaglib/tree/master/src/rnaglib/tasks/RBP_Node)
* [Chemical Modification](https://github.com/cgoliver/rnaglib/tree/master/src/rnaglib/tasks/RBP_Node)
* [Inverse Folding](https://github.com/cgoliver/rnaglib/tree/master/src/rnaglib/tasks/RNA_IF)
* [Small Molecule Ligand Classification](https://github.com/cgoliver/rnaglib/tree/master/src/rnaglib/tasks/RNA_Ligand)
* [Small Molecule Binding Site Detection](https://github.com/cgoliver/rnaglib/tree/master/src/rnaglib/tasks/RNA_Site)
* [RNA Virtual Screening](https://github.com/cgoliver/rnaglib/tree/master/src/rnaglib/tasks/RNA_VS)


See [docs](https://rnaglib.readthedocs.io/en/latest/tasks.html) for more info on usage.

Each link for the tasks above takes you to a `demo.py` file with example usage for each task.

Here is a snippet for a full data loading and model training of the **RNA Binding Site Detection** task. All tasks follow a similar pattern.

```python

from rnaglib.tasks import BindingSiteDetection
from rnaglib.representations import GraphRepresentation

# Load the task data and annotations
ta = BindingSiteDetection('my_root')

# Select a data representation and framework (see docs for support of other data modalities and deep learning frameworks)

ta.dataset.add_representation(GraphRepresentation(framework = 'pyg'))

# Access the predefined splits

train_ind, val_ind, test_ind = ta.split()
train_set = ta.dataset.subset(train_ind)
val_set = ta.dataset.subset(val_ind)
test_set = ta.dataset.subset(test_ind)

```

All you need after this to implement a full training and evaluation loop is standard ML boilerplate. Have a look at one of the [demos](https://github.com/cgoliver/rnaglib/blob/master/src/rnaglib/tasks/RNA_Site/demo.py) for a full example.


### Create your own ML tasks (**new**)

The `rnaglib.tasks` subpackage defines an abstract class to quickly implement machine learning tasks.
You can create a task from the databases available through rnaglib as well as custom annotations to open up your challenge to the rest of the community.
If you implement a task we encourage you to submit a pull request.

To implement a task you must subclass `rnaglib.tasks.Task` and define the following methods:

* `default_splitter()`: a method that takes as input a dataset and returns train, validation, and test indices.
* `build_dataset()`: a method that returns a `rnaglib.RNADataset` object containing the RNAs needed for the task. 

See [here](https://github.com/cgoliver/rnaglib/blob/master/src/rnaglib/tasks/RNA_Site/binding_site.py) for an example of a full custom task implementation.

### Fetch and browse annotated RNA 3D structures

Current release contains annotations generated by x3dna-dssr as well as some additional ones that we added for all available PDBs at the time of release.

Each RNA is stored as a networkx graph where nodes are residues and edges are backbone and base pairing edges.
The networkx graph object has graph-level, node-level and edge-level attributes.
[Here](https://rnaglib.readthedocs.io/en/latest/rna_ref.html) is a reference for all the annotations currently available.

```python

from rnaglib.utils import available_pdbids
from rnaglib.utils import graph_from_pdbid

# load a graph containing annotations from a PDBID
rna = graph_from_pdbid('4v9i')

# you can get list of pdbids currently available in RNAglib
pdbids = available_pdbids()

# print(rna.graph)
{'dbn': {'all_chains': {'num_nts': 143, 'num_chars': 144, 'bseq': 'GCCCGGAUAGCUCAGUCGGUAGAGCAGGGGAUUGAAAAUCCCCGUGUCCUUGGUUCGAUUCCGAGUCUGGGCAC&CGGAUAGCUCAGUCGGUAGAGCAGGGGAUUGAAAAUCCCCGUGUCCUUGGUUCGAUUCCGAGUCCGGGC', 'sstr': '(((((((..((((.....[..)))).(((((.......))))).....(((((..]....))))))))))))..&((((..((((.....[..)))).(((((.......))))).....(.(((..]....))).)))))...', 'form': 'AAAAAA...AA...A.......AAA.AAAA.......A.AAA......AAAAA..A....AAAAAAAAAAAA.-&.AA...AA...A.......AAA.AAAA.......A.AAA......AAAAA..A....A...AAAA.A.-'}...,

```

Databases of annotated structures can be downloaded directly from [Zenodo](https://sandbox.zenodo.org/record/1168342) or through the provided command 
line utility `$ rnaglib_download`.

| Version | Date | Total RNAs | Total Non-Redundant | Non-redundant version | `rnaglib` commit  |
----------|------|------------|---------------------|-----------------------|-------------------|
1.0.0    | 15-02-23 | 5759   | 1176                | 3.269                 |  5446ae2c         |
0.0.0     | 20-07-21 | 3739   | 899                 | 3.186                 |  eb25dabd         |


### Annotate your own structures 

You can extract Leontis-Westhof interactions and convert 3D structures to 2.5D graphs.
We wrap a fork of [fr3d-python](https://github.com/cgoliver/fr3d-python) to support this functionality.

```python
from rnaglib.prepare_data import fr3d_to_graph

G = build_one("../data/structures/1fmn.cif")
```

Warning: this method currently does not support non-standard residues. Support coming soon. Up to version 1.0.0 of the RNA database were created using x3dna-dssr which do contain non-standard residues.

### Quick visualization of 2.5D graphs

We customize networkx graph drawing functionalities to give some convenient visualization of 2.5D base pairing networks.

```python
from rnaglib.drawing import rna_draw
rna_draw(G, show=True, layout="spring")
```

### 2.5D graph comparison and alignment

When dealing with 3D structures as 2.5D graphs we support graph-level comparison through the graph edit distance.

```
from rnaglib.ged import graph_edit_distance
from rnaglib.utils import graph_from_pdbid
G = graph_from_pdbid("4nlf")
print(graph_edit_distance(G, G)) # 0.0
```


## Cite

```
@article{mallet2022rnaglib,
  title={RNAglib: a python package for RNA 2.5 D graphs},
  author={Mallet, Vincent and Oliver, Carlos and Broadbent, Jonathan and Hamilton, William L and Waldisp{\"u}hl, J{\'e}r{\^o}me},
  journal={Bioinformatics},
  volume={38},
  number={5},
  pages={1458--1459},
  year={2022},
  publisher={Oxford University Press}
}
```


### Optional Dependencies

To build 2.5D graphs from scratch locally, for the moment you need to install a fork of `fr3d-python` manually.

```
pip install git+https://github.com/cgoliver/fr3d-python.git
```


## Associated Repositories:

If you use rnaglib in one of your projects, please cite and feel free to make a pull request so we can list your project here.

* [RNAMigos2](https://github.com/cgoliver/RNAmigos2)
* [Structure-and Function-Aware Substitution Matrices](https://github.com/BorgwardtLab/GraphMatchingSubstitutionMatrices)
* [MultiModRLBP: A Deep Learning Approach for RNA-Small Molecule Ligand Binding Site Prediction using Multi-modal features](https://github.com/lennylv/MultiModRLBP)
* [VeRNAl](https://github.com/cgoliver/vernal)
* [RNAMigos](https://github.com/cgoliver/RNAmigos)


## Resources

* [Documentation](https://rnaglib.readthedocs.io/en/latest/?badge=latest)
* [Twitter](https://twitter.com/rnaglib)
* Contact: `rnaglib@cs.mcgill.ca`

## References

1. Leontis, N. B., & Zirbel, C. L. (2012). Nonredundant 3D Structure Datasets for RNA Knowledge Extraction and
   Benchmarking. In RNA 3D Structure Analysis and Prediction N. Leontis & E. Westhof (Eds.), (Vol. 27, pp. 281â€“298).
   Springer Berlin Heidelberg. doi:10.1007/978-3-642-25740-7\_13

